lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs

```diff
@@ -0,0 +1,956 @@
+              , selectionFeeStored = fee selection
+              }
+      Right _ ->
+          Nothing
+
+--------------------------------------------------------------------------------
+-- Selection correctness: fee excess correctness
+--------------------------------------------------------------------------------
+
+data SelectionFeeExcessIncorrectError = SelectionFeeExcessIncorrectError
+    { selectionFeeExcessActual
+        :: Coin
+    , selectionFeeExcessExpected
+        :: Coin
+    }
+    deriving (Eq, Show)
+
+checkFeeExcess
+    :: TxConstraints size
+    -> Selection input size
+    -> Maybe SelectionFeeExcessIncorrectError
+checkFeeExcess constraints selection =
+    checkInner =<< eitherToMaybe (computeCurrentFee selection)
+  where
+    checkInner :: Coin -> Maybe SelectionFeeExcessIncorrectError
+    checkInner currentSelectionFee
+        | selectionFeeExcessExpected == selectionFeeExcessActual =
+            Nothing
+        | otherwise =
+            Just SelectionFeeExcessIncorrectError
+                { selectionFeeExcessActual
+                , selectionFeeExcessExpected
+                }
+      where
+        selectionFeeExcessActual = feeExcess selection
+        selectionFeeExcessExpected = Coin.distance
+            (currentSelectionFee)
+            (computeMinimumFee constraints selection)
+
+--------------------------------------------------------------------------------
+-- Selection correctness: fee sufficiency
+--------------------------------------------------------------------------------
+
+data SelectionFeeInsufficientError = SelectionFeeInsufficientError
+    { selectionFeeActual
+        :: Either NegativeCoin Coin
+    , selectionFeeMinimum
+        :: Coin
+    }
+    deriving (Eq, Show)
+
+checkFeeSufficient
+    :: TxConstraints size
+    -> Selection input size
+    -> Maybe SelectionFeeInsufficientError
+checkFeeSufficient constraints selection =
+    case computeCurrentFee selection of
+        Left nf ->
+            Just SelectionFeeInsufficientError
+                { selectionFeeActual = Left nf
+                , selectionFeeMinimum
+                }
+        Right pf | pf < selectionFeeMinimum ->
+            Just SelectionFeeInsufficientError
+                { selectionFeeActual = Right pf
+                , selectionFeeMinimum
+                }
+        Right _ ->
+            Nothing
+  where
+    selectionFeeMinimum = computeMinimumFee constraints selection
+
+--------------------------------------------------------------------------------
+-- Selection correctness: minimum ada quantities
+--------------------------------------------------------------------------------
+
+data SelectionOutputBelowMinimumAdaQuantityError =
+    SelectionOutputBelowMinimumAdaQuantityError
+        { outputBundle :: TokenBundle
+          -- ^ The output that is below the expected minimum ada quantity.
+        , expectedMinimumAdaQuantity :: Coin
+          -- ^ The expected minimum ada quantity.
```

Comment:
Tbh, wondering if all of this error-boilerplate is worth it. Couldn't `verify` simply be an `Either String ()` in the spec, or something like that?

(Similar to https://github.com/input-output-hk/cardano-wallet/pull/2618/files#r617676467)